src/hardware/vga_xga.cpp

```diff
@@ -873,19 +894,48 @@ void XGA_BlitRect(Bitu val) {
 					break;
 				case 0x02: /* Src is pixel data from PIX_TRANS register */
 					LOG_MSG("XGA: DrawPattern: Wants data from PIX_TRANS register");
+					srcval = 0;
 					break;
 				case 0x03: /* Src is bitmap data */
 					srcval = srcdata;
 					break;
 				default:
 					LOG_MSG("XGA: DrawPattern: Shouldn't be able to get here!");
+					srcval = 0;
 					break;
 			}
+			// For more information, see the "S3 Vision864 Graphics
+			// Accelerator" datasheet
+			//
+			// [http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/Video/VGA/SVGA/S3%20Graphics%2c%20Ltd/S3%20Vision864%20Graphics%20Accelerator%20(1994-10).pdf]
+			//
+			// Page 203 for "Multifunction Control Miscellaneous
+			// Register (MULT_MISC)" which this code holds as
+			// xga.control1, and Page 198 for "Color Compare
+			// Register (COLOR_CMP)" which this code holds as
+			// xga.color_compare.
+
+			auto should_mix_and_draw = [&]() {
+				using namespace bit::literals;
+				// Always update if we're not comparing (COLOR_CMP is bit 8)
+				if (bit::cleared(xga.control1, b8)) {
+					return true;
+				}
+				// If the SRC_NE bit (bit 7) is set ...
+				const auto src_ne_enabled = bit::is(xga.control1, b7);
+				const auto colours_match = (srcval == colorcmpdata);
+				// Then update on a match otherwise update if they differ.
+				return src_ne_enabled == colours_match;
+			};
+			if (should_mix_and_draw()) {
```

Comment:
This works, but a slightly improved pattern for creating a const and initialising it with a lambda would be this. I started using this where appropriate.

```cpp
const auto stuff = [&] {
  ... calculate stuff
}();
```

Reply:
> This works, but a slightly improved pattern for creating a const and initialising it with a lambda would be this. I started using this where appropriate.
> 
> ```c++
> const auto stuff = [&] {
>   ... calculate stuff
> }(); // this calls it immediately
> ```

As far as I know a C++ lambda is a somewhat elaborate function pointer. Are you willing to make a function pointer call per pixel? Depending on architecture that might be a bit of a performance issue.

Reply:
@johnnovak @joncampbell123 -- good points ; I'll aim to drop the lamda.

Reply:
Ok, simplified without the lamba and also avoided using temporary variables.

Reply:
@joncampbell123 Yes, in the very general sense, if you pass lambas around, they become function pointers.

However, they can be trivially inlined in simple cases like this. More recent C++ features don't have to carry the legacy of C ("act as a high level assembler"), and thus have a lot more freedom to perform powerful optimisations.

This is a typical case when I use the lamba to express intent: run this simple code and assign the result to a const. I expect all compilers to inline the lambda. I'm 99% sure theydo, but I will check in godbolt later and report back.

Related:
https://stackoverflow.com/a/13722515

Reply:
Well, I don't have time to prove this is fine @interloper98, and it's a minor thing, so I won't let this hold up the PR process. However, I would bet good money on it it incurs zero performance penalty, so feel free to restore the lambda. Up to you.

Reply:
It's even more straightforward in the current form with fewer intermediate variables, so not much to gain going back to the lamda.

Reply:
Sounds good 😎